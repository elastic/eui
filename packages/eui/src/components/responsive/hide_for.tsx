/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { ReactNode, FunctionComponent } from 'react';
import { useCurrentEuiBreakpoint, EuiBreakpointSize } from '../../services';

export type EuiHideForBreakpoints = EuiBreakpointSize;

export interface EuiHideForProps {
  /**
   * Required otherwise nothing ever gets returned
   */
  children: ReactNode;
  /**
   * List of all the responsive sizes to hide the children for.
   * Array of #EuiBreakpointSize
   */
  sizes: EuiHideForBreakpoints[] | 'all' | 'none';
}

export const EuiHideFor: FunctionComponent<EuiHideForProps> = ({
  children,
  sizes,
}) => {
  const currentBreakpoint = useCurrentEuiBreakpoint();
  const isWithinBreakpointSizes =
    currentBreakpoint && sizes.includes(currentBreakpoint);

  if (sizes === 'all' || isWithinBreakpointSizes) {
    return null;
  }

  return <>{children}</>;
};
