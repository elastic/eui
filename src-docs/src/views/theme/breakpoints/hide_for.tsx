import React from 'react';

import { EuiCode, EuiHideFor, EuiText } from '../../../../../src/components';

export default () => (
  <EuiText>
    <EuiHideFor sizes={'none'}>
      <p>
        Hiding from <EuiCode>{'"none"'}</EuiCode> of the screen sizes
      </p>
    </EuiHideFor>
    <EuiHideFor sizes={['xs']}>
      <p>
        Hiding from <EuiCode>xs</EuiCode> screens only
      </p>
    </EuiHideFor>
    <EuiHideFor sizes={['xs', 's']}>
      <p>
        Hiding from <EuiCode>xs, s</EuiCode> screens
      </p>
    </EuiHideFor>
    <EuiHideFor sizes={['xs', 's', 'm', 'l']}>
      <p>
        Hiding from <EuiCode>xs, s, m, l</EuiCode> screens
      </p>
    </EuiHideFor>
    <EuiHideFor sizes={['xl']}>
      <p>
        Hiding from <EuiCode>xl</EuiCode> screens only
      </p>
    </EuiHideFor>
  </EuiText>
);
