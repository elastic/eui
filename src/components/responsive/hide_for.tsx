/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, {
  ReactNode,
  FunctionComponent,
  useState,
  useEffect,
} from 'react';
import { throttle } from '../../services';
import { EuiBreakpointSize, getBreakpoint } from '../../services/breakpoint';

export type EuiHideForBreakpoints = EuiBreakpointSize;

export interface EuiHideForProps {
  /**
   * Required otherwise nothing ever gets returned
   */
  children: ReactNode;
  /**
   * List of all the responsive sizes to hide the children for.
   * Array of #EuiBreakpointSize
   */
  sizes: EuiHideForBreakpoints[] | 'all' | 'none';
}

export const EuiHideFor: FunctionComponent<EuiHideForProps> = ({
  children,
  sizes,
}) => {
  const [currentBreakpoint, setCurrentBreakpoint] = useState(
    getBreakpoint(typeof window === 'undefined' ? -Infinity : window.innerWidth)
  );

  const functionToCallOnWindowResize = throttle(() => {
    const newBreakpoint = getBreakpoint(window.innerWidth);
    if (newBreakpoint !== currentBreakpoint) {
      setCurrentBreakpoint(newBreakpoint);
    }
    // reacts every 50ms to resize changes and always gets the final update
  }, 50);

  // Add window resize handlers
  useEffect(() => {
    window.addEventListener('resize', functionToCallOnWindowResize);

    return () => {
      window.removeEventListener('resize', functionToCallOnWindowResize);
    };
  }, [functionToCallOnWindowResize]);

  if (
    sizes === 'all' ||
    sizes.includes(currentBreakpoint as EuiBreakpointSize)
  ) {
    return null;
  }

  return <>{children}</>;
};
