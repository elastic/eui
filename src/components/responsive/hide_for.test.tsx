/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from 'enzyme';

import { EuiHideForBreakpoints, EuiHideFor } from './hide_for';

const BREAKPOINTS: EuiHideForBreakpoints[] = ['xs', 's', 'm', 'l', 'xl'];

describe('EuiHideFor', () => {
  // @ts-ignore innerWidth might be read only but we can still override it for the sake of testing
  beforeAll(() => (window.innerWidth = 670));
  afterAll(() => 1024); // reset to jsdom's default

  test('renders', () => {
    const component = render(
      <EuiHideFor sizes={['s']}>
        <span>Child</span>
      </EuiHideFor>
    );

    expect(component).toMatchSnapshot();
  });

  BREAKPOINTS.forEach((size) => {
    test(`${size} is rendered`, () => {
      const component = render(
        <EuiHideFor sizes={[size]}>
          <span>Child</span>
        </EuiHideFor>
      );

      expect(component).toMatchSnapshot();
    });
  });

  test('renders for multiple breakpoints', () => {
    const component = render(
      <EuiHideFor sizes={['s', 'l']}>
        <span>Child</span>
      </EuiHideFor>
    );

    expect(component).toMatchSnapshot();
  });

  test('renders for "none"', () => {
    const component = render(
      <EuiHideFor sizes={'none'}>
        <span>Child</span>
      </EuiHideFor>
    );

    expect(component).toMatchSnapshot();
  });

  test('never renders for "all"', () => {
    const component = render(
      <EuiHideFor sizes={'all'}>
        <span>Child</span>
      </EuiHideFor>
    );

    expect(component).toMatchSnapshot();
  });
});
