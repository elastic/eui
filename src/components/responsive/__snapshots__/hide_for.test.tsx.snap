// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiHideFor l is rendered 1`] = `
<span>
  Child
</span>
`;

exports[`EuiHideFor m is rendered 1`] = `
<span>
  Child
</span>
`;

exports[`EuiHideFor never renders for "all" 1`] = `null`;

exports[`EuiHideFor renders 1`] = `null`;

exports[`EuiHideFor renders for "none" 1`] = `
<span>
  Child
</span>
`;

exports[`EuiHideFor renders for multiple breakpoints 1`] = `null`;

exports[`EuiHideFor s is rendered 1`] = `null`;

exports[`EuiHideFor xl is rendered 1`] = `
<span>
  Child
</span>
`;

exports[`EuiHideFor xs is rendered 1`] = `
<span>
  Child
</span>
`;
